File: solutions/javascript/matching-brackets/1/matching-brackets.js
/**
 * Checks if all brackets in the input string are properly paired.
 *
 * This function removes all non-bracket characters from the input string,
 * then verifies that every opening bracket has a corresponding closing bracket
 * and that they are properly nested.
 *
 * @param {string} input - The input string to check for paired brackets.
 * @returns {boolean} - Returns true if all brackets are correctly paired, otherwise false.
 */
export const isPaired = (input) => {
  const brackets = input.replace(/[^{(\[\])}]/g, ''); //eslint-disable-line

  const bracketsAreMatching = (leftBracket, rightBracket) =>
    (leftBracket === '(' && rightBracket === ')') ||
    (leftBracket === '[' && rightBracket === ']') ||
    (leftBracket === '{' && rightBracket === '}');

  const stack = [];

  for (const char of brackets) {
    if (stack.length && bracketsAreMatching(stack[stack.length - 1], char)) {
      stack.pop();
    } else {
      stack.push(char);
    }
  }

  return stack.length === 0;
};

File: solutions/go/matching-brackets/1/matching_brackets.go
package brackets

var matchingBrackets = map[rune]rune{
	')': '(',
	'}': '{',
	']': '[',
}

// Bracket verifies that any and all pairs of a given string containing brackets `[]`, braces `{}`,
// parentheses `()`, or any combination thereof, are matched correctly.
func Bracket(input string) bool {
	brackets := make([]rune, 0)
	for _, bracket := range input {
		switch bracket {
		case '(', '{', '[':
			brackets = append(brackets, bracket)
		case ')', '}', ']':
			if len(brackets) == 0 || brackets[len(brackets)-1] != matchingBrackets[bracket] {
				return false
			}
			brackets = brackets[:len(brackets)-1]
		}
	}
	return len(brackets) == 0
}
